smarts/core/mission_planner.py

```diff
@@ -182,3 +192,88 @@ def _edge_ids(self, pose: Pose, lane_id: str = None):
             edge_ids.append(next_edges[0].getID())
 
         return edge_ids
+
+    def uturn_waypoints(self, pose: Pose):
+        # TODO: need to revisit the approach to calculate the U-Turn trajectory
+        start_lane = self._road_network.nearest_lane(
+            self._mission.start.position,
+            include_junctions=False,
+            include_special=False,
+        )
+        start_edge = self._road_network.road_edge_data_for_lane_id(start_lane.getID())
+        wp = self._waypoints.closest_waypoint(pose.position)
+        current_edge = self._road_network.edge_by_lane_id(wp.lane_id)
+        if not start_edge.oncoming_edges:
+            return []
+
+        oncoming_edge = start_edge.oncoming_edges[0]
+        oncoming_lanes = oncoming_edge.getLanes()
+        target_lane_index = self._mission.task.target_lane_index
+        if len(oncoming_lanes) <= target_lane_index:
+            return []
+        target_lane = oncoming_lanes[target_lane_index]
+
+        offset = self._road_network.offset_into_lane(start_lane, pose.position[:2])
+        oncoming_offset = max(0, target_lane.getLength() - offset)
+        paths = self.paths_of_lane_at(target_lane, oncoming_offset, lookahead=30)
+        target = paths[0][-1]
+
+        heading = pose.heading
+        target_heading = target.heading
+        lane_width = target_lane.getWidth()
+        lanes = (len(current_edge.getLanes())) + (
+            len(oncoming_lanes) - target_lane_index
+        )
+
+        if current_edge.getID() != oncoming_edge.getID():
+            # agent at the start edge
```

Comment:
I think one reason that the vehicle has some offset during the second part of the uturn is related to the following figure(4 is the end target position and 1 is the vehicle position):
![Screenshot from 2020-11-23 23-15-35](https://user-images.githubusercontent.com/27947083/100050477-7fd08780-2de7-11eb-9b50-1773e8960a98.png)
I think if we force the 3 point to be located on the center of the lane aligned with 4 then we can ensure smoother transition to the end, as @davidrusu mentioned we can work on this in the next PRs.

Reply:
Very nice PR, LGTM.

Reply:
Thanks, @iman512003. You're right, I was fighting with the position of point 3 yesterday but couldn't get it on the right position.